tony-core/src/main/java/com/linkedin/tony/NettyConnection.java

```diff
@@ -0,0 +1,161 @@
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import org.apache.commons.lang3.Range;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * This class encapsulates netty objects related to an established connection which
+ * allows SO_REUSEPORT. It only works with Linux platform since EpollEventLoopGroup used
+ * in {@link NettyConnection#create(int)} is not supported via other platforms. See
+ * <a href="https://netty.io/4.0/api/io/netty/channel/epoll/EpollEventLoopGroup.html">
+ *   https://netty.io/4.0/api/io/netty/channel/epoll/EpollEventLoopGroup.html</a>.
+ */
+final class NettyConnection extends Connection {
+  private static final Log LOG = LogFactory.getLog(NettyConnection.class);
+  final EventLoopGroup eventLoopGroup;
+  final ChannelFuture future;
+  private NettyConnection(EventLoopGroup loopGroup, ChannelFuture future) {
+    this.eventLoopGroup = loopGroup;
+    this.future = future;
+  }
+
+  /**
+   * Closes the netty connection
+   */
+  @Override
+  public void close() {
+    if (this.future.channel().isOpen()) {
+      this.future.channel().close().awaitUninterruptibly();
+    }
+
+    if (!this.eventLoopGroup.isShutdown()) {
+      this.eventLoopGroup.shutdownGracefully().awaitUninterruptibly();
+    }
+  }
+
+  /**
+   * @return the binding port associated with the connection
+   */
+  @Override
+  public int getPort() {
+    InetSocketAddress socketAddress =
+        (InetSocketAddress) this.future.channel().localAddress();
+    return socketAddress.getPort();
+  }
+
+  private static boolean isPortAvailable(int port) throws IOException {
+    ServerSocket serverSocket = null;
+    try {
+      serverSocket = new ServerSocket(port);
+      return true;
+    } catch (Exception e) {
+      return false;
+    } finally {
+      if (serverSocket != null) {
+        serverSocket.close();
+      }
+    }
+  }
+
+  /**
+   * Creates a connection binding to an available port with SO_REUSEPORT.
+   *  See <a href="https://lwn.net/Articles/542629/">https://lwn.net/Articles/542629/</a>
+   *  about SO_REUSEPORT.
+   * @return the created connection
+   */
+  static NettyConnection create() {
+    // Why not using port 0 which lets kernel pick an available port?
+    // - Since another tony executor can bind to the same port when port 0 and SO_REUSEPORT are
+    //   used together. See how a port is selected by kernel based on a free-list and socket
+    //   options: https://idea.popcount.org/2014-04-03-bind-before-connect/#port-allocation.
+    final Range<Integer> portRange = Range.between(1024, 65535);
+    NettyConnection nettyConnection = null;
+    for (int port = portRange.getMinimum(); port <= portRange.getMaximum(); port++) {
+      try {
+        if (isPortAvailable(port)) {
+          nettyConnection = create(port);
+          break;
+        }
+      } catch (Exception exception) {
+        LOG.debug("Fail to create connection to port " + port);
+      }
```

Comment:
thanks,  what you are suggesting here using port zero w/o socket options and find a free port is better. updated the code.

Reply:
Just wanted to note, if you find weird edge cases during testing or in prod:
- SO_REUSEPORT approach still has a drawback that multiple tf processes can bind to the same port, in case of a race condition and compromise integrity of the process. Should be rare.

What we did in model serving for reference:
- Bind to 0 in parent process and find a free port.
- Release it.
- Bind to that port in TF child process.
- Retry a few times if TF child process fails to start-up. **<-- this is the key; retrying makes the probability of failure negligible; and doesn't impact integrity of the process**

Reply:
`SO_REUSEPORT approach still has a drawback that multiple tf processes can bind to the same port, in case of a race condition and compromise integrity of the process. Should be rare.`
can you elaborate more on this? when would this happen?

Reply:
This is the scenario I'm talking about.

Timeline: T1, T2, ..., T5
Executor1:
T1 > Bind to a free port p and release.
T3 > Bind to p with SO_REUSEPORT.
T5 > Child process1 bind to p and listen with SO_REUSEPORT.

Executor2:
T2 > (T3 hasn't happened yet) Bind to a free port p2 == p and release. Probability is very low.
T4 > Bind to p2 == p with SO_REUSEPORT.
T6 > Child process2 bind to p2 == p and listen with SO_REUSEPORT.

Now you have child processes bound to the same port acting as one unit, which is comprising its original intent.

Reply:
thanks, in this case. Would TF child process fail to start up, given both are using SO_REUSEPORT?

Reply:
It would not fail to startup. 
But process A might serve requests meant for process B and can lead to unexpected failures.

Reply:
thanks, i added the corresponding comments to capture this.